Core/UsageSegmentation.swift

```diff
@@ -0,0 +1,100 @@
+//
+//  UsageSegmentation.swift
+//  DuckDuckGo
+//
+//  Copyright © 2024 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+enum UsageActivityType: String {
+
+    case search
+    case appUse = "app_use"
+
+}
+
+protocol UsageSegmenting {
+
+    func processATB(_ atb: Atb, withInstallAtb installAtb: Atb, andActivityType activityType: UsageActivityType)
+
+}
+
+class UsageSegmentation: UsageSegmenting {
+
+    private let pixelFiring: PixelFiring.Type
+    private let storage: UsageSegmentationStoring
+    private let calculatorFactory: UsageSegmentationCalculatorMaking
+
+    init(pixelFiring: PixelFiring.Type = Pixel.self,
+         storage: UsageSegmentationStoring = UsageSegmentationStorage(),
+         calculatorFactory: UsageSegmentationCalculatorMaking = DefaultCalculatorFactory()) {
+        self.pixelFiring = pixelFiring
+        self.storage = storage
+        self.calculatorFactory = calculatorFactory
+    }
+
+    func processATB(_ atb: Atb, withInstallAtb installAtb: Atb, andActivityType activityType: UsageActivityType) {
+        var atbs = activityType.atbsFromStorage(storage)
+
+        // Fail fast by looking at the end of the list
+        guard !atbs.reversed().contains(where: { $0 == atb }) else { return }
```

Comment:
Unsure how `contains` is implemented internally, but maybe the `reversed` here is not needed?

Reply:
From just seeing when the `where` block gets called, it appears to start at the beginning and work forward which I guess normally would be fine, but reversing it means it'll fail faster for atbs already in the list.  tbh in the grand scheme of things probably doesn't make much difference from a performance perspective (already managed to optimise this down quite a bit elsewhere).  Will change.

Reply:
Done